### PR TITLE
LPD-35068

### PR DESCRIPTION
**Pull request by SelenaAungst**
https://liferay.atlassian.net/browse/LPD-35068
Revised version of https://github.com/liferay-platform-experience/liferay-portal/pull/697

### CODE REVIEW

**Comment by SelenaAungst** [comment]
ci:test:sf

**Comment by SelenaAungst** [comment]
ci:test:relevant

**Review by thiago-buarqque — CHANGES_REQUESTED** [review]
Hey @SelenaAungst you PR is good, but it's missing some small stuff.

## Updating services' methods

The relevant CI suite is complaining about the semantic versioning of the API module. That happens because it seems you manually changed the API methods to include the `themeId` and `themeName`. We do not do that manually. When you want a new method or update an existing one, you:

1. Go to the `*ServiceImpl` and/or `*LocalServiceImpl` and update the methods parameters;
2. Commit the changes;
3. Run `buildService` and commit the generated changes.

After that, the API will have the changes based on the impl classes and will update the semantic version of the module.

Let me know if I wasn't clear enough in this part here.

[Here are the classes](https://help.liferay.com/hc/en-us/articles/360029028211-Introduction-to-Service-Builder#:~:text=These%20classes%20are%20where%20you%20implement%20custom%20business%20logic.%20They%E2%80%99re%20the%20only%20classes%20generated%20by%20Service%20Builder%20intended%20for%20customization.) available for customization when working with services.

An example of what I said is in this page [here](https://help.liferay.com/hc/en-us/articles/360030958951-Implementing-an-Add-Method). [Here](https://help.liferay.com/hc/en-us/articles/360030958951-Implementing-an-Add-Method#:~:text=Run%20Service%20Builder%20to%20propagate%20your%20new%20service%20method%20to%20the) is the statement where it says when to run to buildService command again

**Comment by SelenaAungst** [comment]
Thanks @thiago-buarqque! I'll work on making these changes today

**Comment by SelenaAungst** [comment]
I also just saw your comment about updating themeId https://liferay.atlassian.net/browse/LPD-30360 is this something that should be included in this ticket?

**Comment by thiago-buarqque** [comment]
@SelenaAungst No, that is something for a [future story](https://liferay.atlassian.net/browse/LPD-30355).

**Comment by SelenaAungst** [comment]
@thiago-buarqque I tried just updating those files and I am still getting that bnd error when I try to formatSource, is there anything else I need to add for this?

**Comment by SelenaAungst** [comment]
ci:test:sf

**Comment by SelenaAungst** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
Hey @SelenaAungst,

1. You are missing an empty line before the `# breaking` message. It should be:
```
LPD-35068 buildService

# breaking
// ...
```

2. Your style-book-api module is missing the version increase. Run `gw baseline` in that module and commit the changed files in a single commit with the message "LPD-35068 baseline";

3. CI is failing because the columns you added cannot be found after an upgrade is performed. In other words, you are missing the upgrade steps that adds columns to the database for someone coming from a previous version. Please, refer to [this documentation](https://liferay.atlassian.net/wiki/spaces/ENGCOREUPGRADE/pages/1654719711/Upgrade+Development+Training#Schema-version) regarding upgrade processes.

The upgrade process will be something like:
```
@Override
protected void doUpgrade() throws Exception {
    alterTableAddColumn("StyleBookEntry", "themeId", "VARCHAR(255) null");
    alterTableAddColumn("StyleBookEntry", "themeName", "VARCHAR(255) null");
}
```

Check other classes named `*UpgradeProcess` to see examples.

**Comment by SelenaAungst** [comment]
It looks like some of the failures are related to my changes, I will look into this today

**Comment by SelenaAungst** [comment]
I am still seeing some errors from these changes in the portal, I will look into these tomorrow. 

The first is when navigating to the Style Book page this error shows up in the console log:
render.tsx:91 NotFoundError: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
    at is (react-dom.production.min.js:205:62)
    at xc (react-dom.production.min.js:249:52)
    at S.unstable_runWithPriority (scheduler.production.min.js:18:437)
    at Ft (react-dom.production.min.js:120:325)
    at Me (react-dom.production.min.js:244:437)
    at di (react-dom.production.min.js:223:420)
    at Be (react-dom.production.min.js:214:137)
    at Lr (react-dom.production.min.js:279:43)
    at nr (react-dom.production.min.js:282:150)
    at react-dom.production.min.js:287:194

The second error is with Layout, the title in the sidebar shows up as 				"javax.portlet.title.com_liferay_layout_portlets_web_internal_portlet_LayoutPortletsPortlet"
 instead of "Layout".

**Comment by SelenaAungst** [comment]
ci:test:sf

**Comment by SelenaAungst** [comment]
ci:test:relevant

**Comment by algobob** [comment]
> I am still seeing some errors from these changes in the portal, I will look into these tomorrow.
> 
> The first is when navigating to the Style Book page this error shows up in the console log: render.tsx:91 NotFoundError: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node. at is (react-dom.production.min.js:205:62) at xc (react-dom.production.min.js:249:52) at S.unstable_runWithPriority (scheduler.production.min.js:18:437) at Ft (react-dom.production.min.js:120:325) at Me (react-dom.production.min.js:244:437) at di (react-dom.production.min.js:223:420) at Be (react-dom.production.min.js:214:137) at Lr (react-dom.production.min.js:279:43) at nr (react-dom.production.min.js:282:150) at react-dom.production.min.js:287:194
> 
> The second error is with Layout, the title in the sidebar shows up as "javax.portlet.title.com_liferay_layout_portlets_web_internal_portlet_LayoutPortletsPortlet" instead of "Layout".

Could you go in Gogo Shell and run `ds:unsatisfied` ? It will show if some component got broken

**Comment by SelenaAungst** [comment]
> Could you go in Gogo Shell and run `ds:unsatisfied` ? It will show if some component got broken
@algobob I ran this and the output was blank

**Comment by SelenaAungst** [comment]
ci:test:sf

**Comment by SelenaAungst** [comment]
ci:test:relevant

**Comment by SelenaAungst** [comment]
I resolved the issues within style-book, but I am not sure if the layout issue is resolved. I only modified test files for layout as well so maybe the issue is unrelated? @algobob @thiago-buarqque if you test this locally with my changes do you see the issue I mentioned for layout as well?

**Comment by ethib137** [comment]
Hey @SelenaAungst , @algobob , @thiago-buarqque ... I think we need this behind a dev feature flag right?

**Comment by algobob** [comment]
> I resolved the issues within style-book, but I am not sure if the layout issue is resolved. I only modified test files for layout as well so maybe the issue is unrelated? @algobob @thiago-buarqque if you test this locally with my changes do you see the issue I mentioned for layout as well?

ok, I'll try here as well

**Comment by algobob** [comment]
> Hey @SelenaAungst , @algobob , @thiago-buarqque ... I think we need this behind a dev feature flag right?

@ethib137 Maybe we just need one on the frontend, right? The new method on the backend will be used only for style books for now

**Comment by thiago-buarqque** [comment]
I don't think we should only add the FF on the Frontend. The import happen on the backend, so that's a case to check the FF

**Comment by SelenaAungst** [comment]
I will work on adding this, the feature flag should be for LPD-30360, not LPD-35068 right? @thiago-buarqque

**Comment by thiago-buarqque** [comment]
@SelenaAungst Exactly, we use the epic's id since other from that epic will use that feature flag as well

**Comment by ethib137** [comment]
This is looking really good @SelenaAungst ! Thanks for the work everyone. 

Sorry to add a change this late, but as I reviewed this, it still wasn't sitting right that we are adding the themeName to the style book table. The only reason we need this is to show the name when the theme has been removed which makes the style book inactive. I [reached out to Marcos regarding changing the design](https://www.figma.com/design/4oeQva5ooVkUyAaAgfBQmY?node-id=0-1#908842261) to just use the themeId when the theme has been removed. He is fine with that change and has already [updated the figma](https://www.figma.com/design/4oeQva5ooVkUyAaAgfBQmY/LPD-30360?node-id=32-5576&node-type=frame&t=P4fS6O1mx2jM8oCH-0). So, we should update this PR removing themeName and only adding themeId to the table. Does that make sense? Does that sound okay?

cc: @thiago-buarqque @algobob

**Comment by thiago-buarqque** [comment]
@ethib137 I think that's fine... If there isn't a concern regarding the UX, for me, it's okay. And I do understand the argument of having the theme name in the style book entry table. Theme id seems fine, but theme name, yeah, one could argue against it

**Comment by thiago-buarqque** [comment]
Anderson pointed something I forgot about, UI client extensions :sweat_smile:. We discussed and tried it manually and yeah, those don't have a readable id (ERC).

**Comment by SelenaAungst** [comment]
I think also the name should be displayed in the dropdown when creating a new style book, correct? That part still looks the same on the Figma to me

**Comment by ethib137** [comment]
> Actually, I think we need the theme name column. Using only themeId makes sense from workspace flow perspective, but CET deployed from UI generate a uuid as id, which is ilegible.

That's a good point, but I'm not sure if it's enough to warrant adding a column to a table. If we add this column, we then have to manage all the logic for keeping it updated. Do we update the name column every time someone updates the CX through the UI? We could but that's a pain. What about every time a CX or a theme is deployed. The alternative is updating the name column every time the stylebook is saved, by checking what the current name of the theme is. What about for importing a style book? In this case, is the name of the theme included in the imported file? It probably should be, since if the theme wasn't available in Liferay we would need to have it for displaying. 

Basically, I'm concerned with having to manage keeping a column in our table in sync with a column in another table. 

> I think also the name should be displayed in the dropdown when creating a new style book, correct? That part still looks the same on the Figma to me

Yes, and that would still work the same way. In every other situation the theme is still available on the system so we would simply use the themeId to get the theme and return the themeName for displaying. This ensures we always have the current name.

What do you all think?

**Comment by thiago-buarqque** [comment]
For workspace themes (either OSGI or CX) we are fine because their id is similar to the name

**Comment by thiago-buarqque** [comment]
So, if there isn't a concern from the UX regarding showing a ERC to the user, that's fine. But in my opinion, that's not ideal...

**Comment by ethib137** [comment]
Yeah, showing the ERC for CX's added through the UI isn't great, but neither is showing the name since there's no way to add it back and have the style book work. Also, the UI for client extensions is a bad idea in general. It should really only be used for prototyping since a production environment should use the Liferay workspace in order to follow a proper devops workflow. 

So, I really don't think it's worth it. What do you think? @thiago-buarqque are you convinced?

**Comment by thiago-buarqque** [comment]
hmm Okay then, let's go with themeId and create a way (if there isn't one yet) to get the theme name based on the id

**Comment by SelenaAungst** [comment]
Sounds good to me, I will create a new PR for this so it is easier for me to keep track of. I will also add the feature flag and hopefully the layout test issues will be easier to resolve as well.

**Comment by SelenaAungst** [comment]
I made a new PR for these changes here: https://github.com/liferay-platform-experience/liferay-portal/pull/730
